### PR TITLE
Use ct-form-layout class in storage page instead of tables

### PR DESCRIPTION
**Pull request by KKoukiou**
@garrett only the checkboxes in role groups are problematic here I think

### CODE REVIEW

**Comment by KKoukiou** [comment]
testMounting is the only related failure now https://fedorapeople.org/groups/cockpit/logs/pull-11035-20190128-133636-d1218a43-verify-rhel-7-6-distropkg/log.html#242

Looks like the expander does not get triggered by the tests, looks like it's start in some weird state.

**Comment by Gundersanne** [comment]
added `needs-rebase` label, the task bots need `bots/tests-invoke` from current master to work properly.

sorry for the inconvience

**Comment by mvollmer** [comment]
The `SelectSpaces` and `SelectSpace` field types look significantly different now.

New:
image

Old:
image

There is a lot more vertical white space, and the block device name (`/dev/sdb`) is no longer right aligned.

This also makes an old bug much worse: The whole row lights up when hovering, but clicks are only recognized on the checkbox and on the label itself.

**Comment by mvollmer** [comment]
The "checkbox with text" text input doesn't extend all the way to the right anymore:

image

Also, when the dialog is so narrow that the labels have their own rows, the text input should get its own row, I'd say:

image

File: pkg/storaged/dialog.jsx [hunk 1]
```diff
@@ -692,26 +714,67 @@ export const CheckBox = (tag, title, options) => {
         initial_value: options.value || false,
 
         render: (val, change) => {
+            return <CheckBoxComponent tag={tag} val={val} title={title} options={options} update_function={change} />;
+        }
+    };
+};
+
+export const FieldSet = (tag, title, options) => {
```
Comment (mvollmer) [inline pkg/storaged/dialog.jsx]: Let's call this `CheckBoxes` and make `type == "checkbox"` the default.  Checkboxes is all we have for now, and I think it's an important special case to give it its own name.

File: pkg/storaged/dialog.jsx [hunk 2]
```diff
@@ -684,6 +687,25 @@ export const SelectSpace = (tag, title, options) => {
...
+                <a className="dialog-item-tooltip">
+                    <span className="fa fa-lg fa-info-circle" />
+                </a>
+            </OverlayTrigger>
+            }
+        </div>
+    );
+};
+
 export const CheckBox = (tag, title, options) => {
```
Comment (mvollmer) [inline pkg/storaged/dialog.jsx]: Let's remove `CheckBox` altogether.  The remaining uses are "Store passphrase" in format-dialog.jsx, and the VDO options.  The latter need to be a field set anyway.

File: pkg/storaged/dialog.jsx [hunk 3]
```diff
@@ -692,26 +714,67 @@ export const CheckBox = (tag, title, options) => {
...
-                    }
+                <div role="group" className="ct-form-layout-vertical">
+                    { fieldset }
                 </div>
             );
         }
     };
 };
 
+const TextInputCheckedComponent = ({ tag, val, title, update_function }) => {
```
Comment (mvollmer) [inline pkg/storaged/dialog.jsx]: Let's remove TextInputChecked altogether.  There are no uses anymore.

**Comment by mvollmer** [comment]
For reference, this is how to get a reliable failure fom TestStorage.testMounting:

```diff
diff --git a/pkg/lib/listing.less b/pkg/lib/listing.less
index 9c056c97b..ee8ed9f7f 100644
--- a/pkg/lib/listing.less
+++ b/pkg/lib/listing.less
@@ -171,7 +171,7 @@ td.listing-ct-toggle i:before {
 
 tbody.open td.listing-ct-toggle i:before {
     content: "\f107";
-    animation: listingCtToggleOpen 100ms ease-in-out;
+    animation: listingCtToggleOpen 2000ms ease-in-out;
     top: 1px;
 }
 
@@ -455,7 +455,7 @@ tbody.open > tr.listing-ct-body {
 /* (Switching tabs will adds new listing-ct-bodys & toggle visibility) */
 tbody.open div.listing-ct-head,
 tbody.open div.listing-ct-head + div.listing-ct-body:not([hidden]):last-child {
-    animation: listingCtShow 100ms ease-in-out;
+    animation: listingCtShow 2000ms ease-in-out;
 }
 
 @keyframes listingCtShow {

```

File: test/verify/storagelib.py [hunk 4]
```diff
@@ -172,11 +172,11 @@ def teardown():
 
     def content_tab_info_row(self, row_index, tab_index, title):
```
Comment (mvollmer) [inline test/verify/storagelib.py]: Let's rename this this to `content_tab_info_label` since it no longer refers to an element that includes the whole row.

**Comment by mvollmer** [comment]
The `tab-row-actions` buttons are misplaced:
 
image

Here, the "Mount" button should be in the same row as "Mount Point".

File: test/common/test-functions.js [hunk 5]
```diff
@@ -99,7 +99,8 @@ function ph_mouse(sel, type, x, y, btn, force) {
     let el = ph_find(sel);
 
     /* The element has to be visible, and not collapsed */
-    if (!force && (el.offsetWidth <= 0 || el.offsetHeight <= 0))
+    console.log("v?", sel, el.offsetWidth || "-", el.offsetHeight || "-")
```
Comment (mvollmer) [inline test/common/test-functions.js]: This line should be removed, obviously.

File: test/common/test-functions.js [hunk 6]
```diff
@@ -150,6 +151,7 @@ function ph_set_checked (sel, val)
 function ph_is_visible (sel)
 {
     var el = ph_find(sel);
+    console.log("V?", sel, el.offsetWidth || "-", el.offsetHeight || "-")
```
Comment (mvollmer) [inline test/common/test-functions.js]: This line should be removed, obviously.

Comment (KKoukiou) [reply test/common/test-functions.js]: Done

File: test/verify/storagelib.py (hunk 4 above)
Comment (KKoukiou) [reply test/verify/storagelib.py]: fixed

File: pkg/storaged/nfs-details.jsx [hunk 7]
```diff
@@ -169,14 +169,19 @@ export function nfs_fstab_dialog(client, entry) {
...
-                                             }),
-                                    CheckBox("mount_ro", _("Mount read only"),
-                                             { value: opt_ro }),
-                                    TextInputChecked("mount_extra_options", _("Custom mount option"),
-                                                     { value: extra_options === "" ? false : extra_options })
+                                    FieldSet("mount_options", _("Mount Options"),
+                                             { fields: [
+                                                 { title: _("Mount at boot"), tag: "auto", type: "checkbox" },
+                                                 { title: _("Mount read only"), tag: "ro", type: "checkbox" },
+                                                 { title: _("Mount extra options"), tag: "extra", type: "checkboxWithInput" },
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/storaged/nfs-details.jsx]: fixed

File: pkg/storaged/format-dialog.jsx [hunk 8]
```diff
@@ -82,36 +82,30 @@ export function mounting_dialog_fields(is_custom, mount_dir, mount_options, visi
...
+                 { visible: function (vals) {
+                     return visible(vals) && vals.mounting == "custom";
+                 },
+                   value: { auto: opt_auto,
+                            ro: opt_ro,
+                            extra: extra_options === "" ? false : extra_options },
+                   fields: [
+                       { title: _("Mount at boot"), tag: "auto", type: "checkbox" },
+                       { title: _("Mount read only"), tag: "ro", type: "checkbox" },
+                       { title: _("Mount extra options"), tag: "extra", type: "checkboxWithInput" },
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/storaged/format-dialog.jsx]: fixed

**Comment by KKoukiou** [comment]
> The `tab-row-actions` buttons are misplaced:
> 
> image
> 
> Here, the "Mount" button should be in the same row as "Mount Point".

Fixed this specific one. Let me know if you notice another faulty one.

**Comment by KKoukiou** [comment]
> `SelectSpaces` and `SelectSpace`

Good point. @garrett I think this is the new css causing this here. We have the same appearance in networking. Should be fix it everywhere?

**Comment by KKoukiou** [comment]
> The "checkbox with text" text input doesn't extend all the way to the right anymore:
> 
> image
> 
> Also, when the dialog is so narrow that the labels have their own rows, the text input should get its own row, I'd say:
> 
> image

I think @garrett will be able to give some good feedback for this as well.

**Comment by mvollmer** [comment]
> Fixed this specific one. Let me know if you notice another faulty one.

I think this was actually the only faulty one.

**Comment by mvollmer** [comment]
Rebased on new master.

File: pkg/storaged/fsys-tab.jsx [hunk 9]
```diff
@@ -149,71 +149,50 @@ export class FilesystemTab extends React.Component {
...
+                        </React.Fragment>
+                    )}
+
+                    { (mounted_at.length > 0 || self.props.client.is_old_udisks2) && (
+                        <React.Fragment>
+                            <label className="control-label">{_("Mounted At")}</label>
+                            <div>
+                                {mounted_at.join(", ")}
+                                <div className="tab-row-actions">
+                                    <StorageButton onClick={unmount}>{_("Unmount")}</StorageButton>
```
Comment (mvollmer) [inline pkg/storaged/fsys-tab.jsx]: This needs to offer "Mount" when mounted_at.length == 0, see the old code.

Comment (KKoukiou) [reply pkg/storaged/fsys-tab.jsx]: Ah sorry, my fault. Let's remove this commit then.

**Comment by mvollmer** [comment]
> The SelectSpaces and SelectSpace field types look significantly different now.

Fixed now with some new flexbox CSS.

**Comment by mvollmer** [comment]
> The "checkbox with text" text input doesn't extend all the way to the right anymore:

This is too hard for me to fix.  @garret?

File: pkg/storaged/fsys-tab.jsx (hunk 9 above)
Comment (mvollmer) [reply pkg/storaged/fsys-tab.jsx]: Fixed.

**Comment by garrett** [comment]
I have some additional fixes in #11154. I'll merge over here to test it.

**Comment by garrett** [comment]
OK, so the changes in #11154 and the changes here don't play well together. I'll have to work on this a bit more to make everything happy. :unamused:

**Comment by garrett** [comment]
(BTW: @mvollmer: you dropped a `t` again.)

**Comment by KKoukiou** [comment]
@mvollmer It looks great to me, by of course I can't approve it since I created this PR initially. Feel free to approve and let's merge it.
Improvements as adding htmlFor attr to labels can be made in followups, since we didn't cause any regression here.

**Comment by martinpitt** [comment]
This needs a pep8 fix:

    test/verify/storagelib.py:180:27: E225 missing whitespace around operator
    test/verify/storagelib.py:182:27: E225 missing whitespace around operator

See our [HACKING](https://github.com/cockpit-project/cockpit/blob/master/HACKING.md) how to set up a git pre-push hook to avoid this in the future :-)

**Comment by mvollmer** [comment]
> I'll have to work on this a bit more to make everything happy.

@garrett We will merge this PR and come back to the (minor) layout issues later, ok?

**Comment by mvollmer** [comment]
Added "Closes #..."

**Comment by garrett** [comment]
I'm still working on this. :wink:

**Comment by garrett** [comment]
I've pushed up some small React code changes that add `className`s to a few elements. This is detangled from the CSS related changes that I guess I'll move to a follow-up PR (which would probably include Storage-specific changes as a separate commit, in addition to ct-form-layout).

Hopefully this lets us merge this PR and address the layout concerns separately.

**Comment by mvollmer** [comment]
> I've pushed up some small React code changes that add classNames to a few elements. 

Ok, thanks!

**Comment by martinpitt** [comment]
Apparently [udisks crashed on debian-testing](https://fedorapeople.org/groups/cockpit/logs/pull-11035-20190219-190202-1ab7b4d3-verify-debian-testing/log.html#193), retrying.

**Comment by KKoukiou** [comment]
nfs-mount-new-mobile
nfs-mount-new-wide

**Pull request merged**